exec/slicemachine.go

```diff
@@ -189,6 +188,7 @@ loop:
 		select {
 		case <-time.After(statsPollInterval):
 		case <-ctx.Done():
+			return
```

Comment:
This seems a little suspect. I don't think we should assume that context cancelation only comes from session shutdown, and it seems to me there isn't any harm in going through the task tear down code that it currently does?

Reply:
Yeah, that's a good point.

The goal here is to avoid the confusing spurious logging on shut down, i.e. tasks reported lost even though we know all computations have completed successfully. Some other possibilities:
* Have an alternate signal to shutdown, e.g. a separate channel passed in. This seemed messy in its own way, but  it's probably more semantically correct.
* Set some shared state during shut down to signal to not log the lost task messages.
* Just make the logging message less scary, e.g. "machine shutting down; 845 tasks may be recomputed if necessary".

Reply:
I'm not sure what you mean by the current task tear down code. Isn't that the code that emits the log we are trying to avoid?

Reply:
Another possibility is to attach state to the context itself, something like:
```go
func NewContext(ctx, sess) context.Context { ... }
func InShutdown(ctx) bool { ... }
```

@mariusae: Do you have suggestions or thoughts on dealing with these spurious logs? (It's been noted as a point of confusion by a few people, which is why we're trying to do something about it).

Reply:
My vote: make the log messages less scary, and avoid adding the extra complexity of different shutdown signals, etc. I think that's a sufficiently good UX.

Reply:
Maybe it's enough to just make the `bigslice` shutdown synchronous. I think part of the confusion was that people call the shutdown function, scan/print their results or other logging, and then subsequently see a litany of "task lost" logs, obscuring their printed results. @josh-newman: what do you think of that trade-off (basically keep the `WaitGroup` stuff but leave the task lost logging)?

Reply:
I just had a quick look at `bigmachine` shutdown, and it looks like it would be similarly straightforward to make its shutdown synchronous too, which would then create the nice property that you wouldn't get any logging after calling the session shutdown.

(Note that we already have the session take exclusive ownership of the bigmachine system; it already shuts the `bigmachine.B` down which in turn shuts down the underlying system. This would just make that synchronous.)

Reply:
I'd prefer just one codepath here. What if we just remove the log altogether? Instead, we could have a log message that indicates that a task is being recomputed because it was previously lost. (Maybe we do this already?) It seems to me that this message isn't really useful to the user anyway. It only sows confusion. Presumably a truly interested user can use the other monitoring facilities to understand this kind of behavior.

I'm also not opposed to making shutdown synchronous, but that's almost an orthogonal thing.

Reply:
I agree about only having one codepath, so I think we're settled there.

I think the message is useful for very interested users (like me), as it has provided a useful landmark in the logs when tracking a task that has unexpected errors. These are much more rare given all of the improvements that we've made to fault-tolerance and error handling, but I think they still have some use. I think it falls into the large class of messages that are useful if you're debugging internals or confusing behavior but not useful to end users, e.g. logging attempts to start tailing on bootup, logging worker boot, etc.

How's this for a path forward?

* Redouble efforts to hide/classify non-error logs in general, this "lost machine" one being one of them. Do this in a separate set of PRs.
* Transform this PR into a synchronous shutdown PR, as I think it's generally the least surprising behavior. This logging message aside, more than one person has asked me why it we still contact machines after `shutdown`.

Reply:
I agree. Synchronous shutdown seems like a good property regardless.

Reply:
This all sounds good to me, just have one concern about synchronous shutdown. If we just add wait groups like we have now with no channel being used to shutdown the slicemachine.Go goroutines, the synchronous shutdown will last 10 minutes waiting for the KeepAlives from bigmachine to stop. Should we start with a change to bigmachine?

Reply:
I would revert to the previous version that tied the contexts to a shutdown channel, https://github.com/grailbio/bigslice/pull/90/commits/c5e7805b246554b19e23d12fdcf29fc35cafed90 . Instead of returning when the context is done, just do what the existing code does, which is just break the loop.

Reply:
```suggestion
```
Per this thread.